### PR TITLE
fix(category): NEXT-33511 - Fix slot config inheritance for CMS layout overrides via category

### LINKED ISSUE #5655: Implement inheritance for JSON-Blobs
Currently, JSON blobs kill inheritance. When you have a json-block with multiple fields, like in the slotConfig of Categories, as soon as one of those fields are filled, the others lose all inheritance features, since it only replaces the whole blob and not parts of it.

See the linked tickets for a specific example. Please test this with categorie layout overrides.

### PR DESCRIPTION
**Pull request by marcelbrode**
### 1. Why is this change necessary?
As soon as a category overrides anything in the layout, a json-blob with changes will be created. This also won't be cleaned if an element in the CMS has been removed, therefore an overriden block which got deleted afterwars will still leave content. Since the DAL delievers inherited values fieldwise, the describes way will lead to a case where inheritance is broken, since the json-blob has no data of current blocks. This MR fixes this issue

### 2. What does this change do, exactly?
Loads all translations in the inheritance chain (Selected Language -> Language it inherits from, if configured -> System Default language) and merges them properly, to get this structure:

```mermaid
flowchart LR
 subgraph austrian["Austrian (Inherits German)"]
        AT("Template")
        AO("Override")
  end
 subgraph german["German"]
        GT("Template")
        GO("Override")
  end
 subgraph english["System default (English)"]
        ET("Template")
        EO("Override")
  end
    AO --> GO
    GO --> EO
    EO --> AT
    AT --> GT
    GT --> ET
```

### 3. Describe each step to reproduce the issue or behaviour.

- Construct the following language situation: 
  - Have a parent language (german)
  - Have a child language, which inherits from the parent one (austrian)
- Create a CMS page with an element in a fallback language (german)
- Assign it to a Category page
  - Override the element in a parent language (german)
- Get back to CMS and remove the element and create a new one, exactly like before
  - Let the element state it's from the template for easier validation
  - _Now, the category will have data for the slot config to override, but without a target, so nothing to override_ 
- Get back to the Category page and override the new element in the parent (german)
- Visit the page in the storefront in the child language (austrian)
  - Before: You see the parent TEMPLATE (german) instead of its override
  - After: You see the parent override (german), like it is supposed to be

### 4. Please link to the relevant issues (if any).
- closes #5655
- relates #6238

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

### CODE REVIEW

**Comment by mitelg** [comment]
is this a backport? if not, it should be merged to trunk first

**Comment by marcelbrode** [comment]
> is this a backport? if not, it should be merged to trunk first

I forgot to put this into draft mode. Is this possible after creating the PR? Can't find a button for that. Was more like a safety push to test it initially

**Comment by mitelg** [comment]
image

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php [hunk 1]
```diff
@@ -138,4 +139,27 @@ private function createCriteria(string $pageId, Request $request): Criteria
...
+     * @return array<string, array<string, mixed>>|null
+     */
+    private function buildCmsSlotConfig(CategoryEntity $category, SalesChannelContext $context): ?array
+    {
+        $inheritanceChain = $context->getLanguageIdChain();
+        if (\count($inheritanceChain) <= 1) {
+            return $category->getTranslation('slotConfig');
+        }
+
+        $translatedSlotConfigs = \array_map(static function ($languageId) use ($category) {
```
Comment (dgrothaus-sw) [inline src/Core/Content/Category/SalesChannel/CategoryRoute.php]: You could strictly type the `$languageId` in the Lambda.

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php [hunk 2]
```diff
@@ -138,4 +139,27 @@ private function createCriteria(string $pageId, Request $request): Criteria
...
+            /** @var ?CategoryTranslationEntity $currentTranslation */
+            $currentTranslation = \array_find(
+                $category->getTranslations()?->getElements() ?? [],
+                static fn ($translation) => $translation->getLanguageId() === $languageId
```
Comment (dgrothaus-sw) [inline src/Core/Content/Category/SalesChannel/CategoryRoute.php]: Same here. `$translation` could be strictly typed.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php [hunk 3]
```diff
@@ -139,11 +162,128 @@ public function testHomeWithSalesChannelOverride(): void
...
+    /**
+     * @return list<array<string, mixed>>
+     */
+    private function createProducts(): array
+    {
+        $product = [
+            'name' => 'test',
+            'stock' => 10,
+            'price' => [
+                ['currencyId' => Defaults::CURRENCY, 'gross' => 15, 'net' => 10, 'linked' => false],
```
Comment (dgrothaus-sw) [inline tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:265]: It's just a test product, but I somehow feel offended by the fact that a 15% tax on a 10.00 bucks product leads to a gross price increase of 50%.

File: tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php [hunk 4]
```diff
@@ -0,0 +1,798 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Tests\Integration\Core\Content\Category\SalesChannel\fixtures;
+
+use PHPUnit\Framework\Attributes\Group;
+
+/**
+ * @internal
+ *
+ * @phpstan-type CmsInheritanceDataProviderIterator iterable<array{
```
Comment (dgrothaus-sw) [inline tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php]: This is the same as `CmsInheritanceDataProviderIterator` from `CategoryRouteTest`. You can `@phpstan-import-type` the type from the other Test or define it here and import it in the other test file which logically makes more sense IMHO.

File: tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php [hunk 5]
```diff
@@ -0,0 +1,798 @@
...
+                'hasSlotOverride' => [],
+            ],
+            'expected' => 'at Template',
+        ];
+    }
+
+    /**
+     * @return CmsInheritanceDataProviderIterator
+     */
+    public static function allOverridesDataProvider(): iterable
```
Comment (dgrothaus-sw) [inline tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php:147]: Would it be a good idea to add a case with a duplicate of one the languages to test if the `\array_reverse(\array_unique())` from the `CategoryRoute::buildCmsSlotConfig()` works as intended?

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php (hunk 3 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:265]: xD Good point. But this was only moved, so it wasn't my doing. Someone's math is bad

File: tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php (hunk 4 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php]: Didn't know that. Nice catch!

File: tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php (hunk 5 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/fixtures/CategoryRouteInheritanceFixtures.php:147]: Nice idea! Done

**Review by dgrothaus-sw — APPROVED** [review]

**Review by ravenokko — APPROVED** [review]

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php [hunk 6]
```diff
@@ -60,20 +61,19 @@ public function load(string $navigationId, Request $request, SalesChannelContext
 
         $category = $this->loadCategory($navigationId, $context);
 
-        if (($category->getType() === CategoryDefinition::TYPE_FOLDER
-                || $category->getType() === CategoryDefinition::TYPE_LINK)
-            && $context->getSalesChannel()->getNavigationCategoryId() !== $navigationId
-        ) {
+        $categoryHasNonPageType = \in_array($category->getType(), [CategoryDefinition::TYPE_LINK, CategoryDefinition::TYPE_FOLDER], true);
```
Comment (MartinKrzykawski) [inline src/Core/Content/Category/SalesChannel/CategoryRoute.php]: is it not easier to check if the category type is not from type `page`? The current implementation will make problems if we provide new types and we would need to extend this array always if new types have been added.

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php [hunk 7]
```diff
@@ -138,4 +139,27 @@ private function createCriteria(string $pageId, Request $request): Criteria
 
         return $criteria;
     }
+
+    /**
+     * @return array<string, array<string, mixed>>|null
+     */
+    private function buildCmsSlotConfig(CategoryEntity $category, SalesChannelContext $context): ?array
```
Comment (MartinKrzykawski) [inline src/Core/Content/Category/SalesChannel/CategoryRoute.php]: This function includes not covered lines of code, please cover them with unit tests.

Comment (MartinKrzykawski) [inline src/Core/Content/Category/SalesChannel/CategoryRoute.php]: This function is hard to read imo.

this could help:

1. add a comment for each section to explain what it does
2. use more private functions intern with a good naming, then you do not have to follow the whole logic and you understand it directly because the namings explain it enough.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php [hunk 8]
```diff
@@ -2,57 +2,75 @@
...
-            'id' => $this->ids->create('sales-channel'),
-            'navigationCategoryId' => $this->ids->get('home-category'),
-        ]);
-
-        $this->setVisibilities();
     }
 
     public function testCmsPageResolved(): void
     {
+        $this->createListingData();
```
Comment (MartinKrzykawski) [inline tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:60]: why not in the setUp method?

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php [hunk 9]
```diff
@@ -139,11 +162,207 @@ public function testHomeWithSalesChannelOverride(): void
...
+    private function createDomains(Context $context): void
+    {
+        $basicDomainData = [
+            'salesChannelId' => $this->ids->get('sales-channel'),
+            'currencyId' => Defaults::CURRENCY,
+            'snippetSetId' => $this->getSnippetSetIdForLocale('en-GB'),
+        ];
+
+        $url = 'http://localhost:8000';
+        static::getContainer()->get('sales_channel_domain.repository')->create([[
```
Comment (MartinKrzykawski) [inline tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php]: getContainer() is not a static call.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php [hunk 10]
```diff
@@ -139,11 +162,207 @@ public function testHomeWithSalesChannelOverride(): void
...
+    /**
+     * @return CmsInheritanceDataProviderIterator
+     */
+    public static function categoryCmsPageInheritanceDataProvider(): iterable
+    {
+        yield from CategoryRouteInheritanceFixtures::noOverridesDataProvider();
+        yield from CategoryRouteInheritanceFixtures::mixedOverridesDataProvider();
+        yield from CategoryRouteInheritanceFixtures::allOverridesDataProvider();
+
+        yield from CategoryRouteInheritanceFixtures::enOverridesDataProvider();
```
Comment (MartinKrzykawski) [inline tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:180]: are these 3 override checks really required? I mean the logic is the same, only the languages are different.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php [hunk 11]
```diff
@@ -139,11 +162,207 @@ public function testHomeWithSalesChannelOverride(): void
...
+     *  Testing CMS slot inheritance with category overrides, with multiple EN occurances in language chain.
+     *  - EN is System Default language
+     *  - DE inherits from EN
+     *  Expected hierarchy: Overrides of categories (DE > EN) > Templates of categories (DE > EN)
+     *  Actual language chain hierarchy: DE > EN > EN
+     *
+     * @param CmsInheritanceDataProviderActual $actual
+     */
+    #[DataProvider('categoryCmsPageInheritanceWithDuplicatesInLanguageChainDataProvider')]
+    public function testCategoryCmsPageInheritanceWithDuplicatesInLanguageChain(array $actual, string $expectedContentValue): void
```
Comment (MartinKrzykawski) [inline tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php]: You test a lot of cases via integration tests. This can affect our integration runtime duration. Do you checked how long it takes to run these tests? You should use integration tests for the general process and cover all edge cases with unit tests imo.

**Review by MartinKrzykawski — CHANGES_REQUESTED** [review]

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php (hunk 6 above)
Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: Other Category-Types are most likely displaying types. Folder and Link are "dumb" types and I guess it has to be adjusted anyway, you know? That place is just a change for better readability anyway. 

Would you prefer constants for "dumb" and "Non-dumb" types or sth like that?

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php (hunk 8 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:60]: Because the data is depending on the case. My Tests do not work with ListingData, therefore I pushed those to the cases.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php (hunk 10 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:180]: I can remove some of the tests if you think it's too much, but it's not exactly the same, since en would fall back to Template immediately, while de has en in between and so on. I guess I'd rather shrink the amount than leave a complete "category" of cases out of the tests.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php (hunk 11 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php]: Locally like 5 seconds for all 63 cases. But of course, you have point there. I'll have a look how other route tests do unit tests 😄

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php (hunk 6 above)
Comment (MartinKrzykawski) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: category has 3 types: page, link and folder.

My point is, why you do not check it like this: $categoryHasNonPageType = $category->getType() !== CategoryDefinition::TYPE_PAGE

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php (hunk 11 above)
Comment (MartinKrzykawski) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php]: 5 seconds are a lot for a single test class 🤔 we should not forget that these 5 seconds are on your fast machine and not on our github runner. The whole magic appears in the `buildCmsSlotConfig ` function, right? You could write unit tests and cover all edge cases with unit tests. This would be much faster, because you would not insert data to the database and send requests in each single run. For the integration part i would keep only few specific cases if the inheritance works in general :)

Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php]: Sounds good 👍

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php (hunk 6 above)
Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: Because I didn't change the logic, only the readability. But sure, we can do that!

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php (hunk 7 above)
Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: I hope that splitting into multiple functions make it more clear. Please tell me if not!

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php (hunk 6 above)
Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: I reverted the simplification, because there is more than these 3 types like landing_page, product_detail and product_list.

Comment (MartinKrzykawski) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: But they are a different category as the page types or? the constant is `LINK_TYPE_LANDING_PAGE` and not just `TYPE_FOO`

Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: The Subtypes of Link can appear as well, therefore we need to be explicit here.

Comment (MartinKrzykawski) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: but they are not set as type, they are set as linkType - it is an own property in the CategoryEntity

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php [hunk 12]
```diff
@@ -138,4 +139,42 @@ private function createCriteria(string $pageId, Request $request): Criteria
...
     }
+
+    /**
+     * @return array<string, array<string, mixed>>|null
+     */
+    private function buildMergedCmsSlotConfig(CategoryEntity $category, SalesChannelContext $context): ?array
+    {
+        $inheritanceChain = $context->getLanguageIdChain();
+        if (\count($inheritanceChain) <= 1) {
+            return $category->getTranslation('slotConfig');
```
Comment (MartinKrzykawski) [inline src/Core/Content/Category/SalesChannel/CategoryRoute.php:150]: line is still not covered by an unit test

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php [hunk 13]
```diff
@@ -2,57 +2,75 @@
...
+ *  }
+ * @phpstan-type CmsInheritanceDataProviderIterator iterable<array{
+ *       actual: CmsInheritanceDataProviderActual,
+ *       expected: string
+ *  }>
  */
 #[Group('store-api')]
+#[Package('discovery')]
+#[CoversClass(CategoryRoute::class)]
 class CategoryRouteTest extends TestCase
```
Comment (MartinKrzykawski) [inline tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:38]: Your tests include deprecations and log warnings in the pipeline: https://github.com/shopware/shopware/actions/runs/13452960407/job/37590734655?pr=6598

logs

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php (hunk 12 above)
Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php:150]: As I told you before, I didn't check that yet.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php (hunk 11 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php]: Reduced it from to ~3.5s, I'd like to keep the rest of the tests, so that each case is still adequately covered. Hope that's fine 👍

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php [hunk 14]
```diff
@@ -138,4 +139,42 @@ private function createCriteria(string $pageId, Request $request): Criteria
...
+        }
+
+        /** @var non-empty-list<string> $languageMergeOrder */
+        $languageMergeOrder = \array_reverse(\array_unique($inheritanceChain));
+        $translatedSlotConfigs = $this->getTranslatedSlotConfigs($category, $languageMergeOrder);
+
+        return \array_merge(...$translatedSlotConfigs);
```
Comment (MartinKrzykawski) [inline src/Core/Content/Category/SalesChannel/CategoryRoute.php:157]: is the array merge required? getTranslatedSlotConfigs returns just an array, or? 🤔

Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php:157]: Of course, that's the most important part. This function created the inheritance hierarchy of all translations. Otherwise it would be a translation collection. 

getTranslatedSlotConfigs returns an array of config-arrays, which will be merged for actual usage.

File: tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php (hunk 13 above)
Comment (marcelbrode) [reply tests/integration/Core/Content/Category/SalesChannel/CategoryRouteTest.php:38]: Thanks 👍 fixed

File: src/Core/Content/Category/SalesChannel/CategoryRoute.php (hunk 6 above)
Comment (marcelbrode) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: It looks like it was intentional. The whole thing goes into the CategoryUrlGenerator and runs through all the types. It was probably done this way to avoid constantly having to test the combination of two variables that would have to work together all the time anyway. I think it could be improved, but that would cause a breaking change.

Comment (MartinKrzykawski) [reply src/Core/Content/Category/SalesChannel/CategoryRoute.php]: Not sure how you recognize it - imo the service agrees my opinion. But it is okay for me. I had a second thought about it and it might be that other types will be added and they are maybe not required for this evaluation but would be part of it then. Thats okay for me.

**Review by MartinKrzykawski — APPROVED** [review]
